[sensorfwd] Transfer events from worker threads via pipe

spiiroin opened the pull request.

spiiroin committed:
[sensorfwd] Transfer events from worker threads via pipe. Fixes JB#60382

Non-reentrant method HybrisManager::processSample() is used directly
from hal worker thread. This can lead to heap corruption and crashes.

Use pipe and io-watch for transferring sensor events from worker
thread to mainloop context and thus avoid the reentrancy issues
by having sensor chain data management happen only in one thread.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

mlehtima committed:
[sensorfw] Add support for Android sensor device API 1.0. JB#60382

mlehtima commented on core/hybrisadaptor.cpp:
```diff
@@ -1228,59 +1248,152 @@ void *HybrisManager::eventReaderThread(void *aptr)
+    if (wakeupEventCount)
+        ObtainTemporaryWakeLock();
+
+
```
Nitpick, extra empty line

spiiroin replied:
Fixed

jusa commented on core/hybrisadaptor.cpp:
```diff
@@ -1228,59 +1248,152 @@ void *HybrisManager::eventReaderThread(void *aptr)
+        m_eventPipeNotifier->setEnabled(false);
+        sensordLogW("fd mismatch, event pipe notifier disabled");
+    } else {
+        const size_t maxEvents = 64;
```
These two magic values could be somewhere?

spiiroin replied:
Yes, ideally it should be one and same constant.
I started doing that at some stage but gave up as it yielded a lot of diff noise.
Maybe I should do it as a separate commit...

mlehtima replied:
Just a define at the top of the file should be simple enough.

spiiroin replied:
Yeah, but there are other itches to scratch once one starts...

spiiroin committed:
[sensorfwd] Use common constant for event queue size. JB#60382

Number of events to ask from android hal/service should agree
with number of events to transfer over event pipe in one go.
Which they do at the moment, but using separately defined
hard coded values.

Use common maxEvents = 64 constant for both transfer sizes
within sensorfwd and from android hal/service.

Also harmonize naming of event count variables and parameters.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>
